blog/_posts/2017-09-06-gsoc-derivative_operators.md

```diff
@@ -0,0 +1,156 @@
+u_i^0 = u_0(x_i) \textit{ for } i = 1,...,N\\ 
+u_0^j = u_{N+1}^j = 0 \textit{ for } j = 1,...,M+1\\
+\end{matrix}\right.
+$$
+
+This is the stencil rewritten as a recurrence. Writing it out in vector form, we get:-
+$\frac{U^{j+1} - U^j}{k} + A_h U^j = F^j \textit{ for } j = 1,...,M$
+
+When we want to apply this operator on the vector $U$, the weight vector turns into a matrix called the transformation matrix $A_h$
+
+$$ A_h = \begin{pmatrix}
+2 & -1 & 0 &  \cdots  & 0\\ 
+-1 & 2 & -1 &  \cdots  & 0\\ 
+\vdots & \ddots & \ddots & \ddots & \vdots \\ 
+0 & \cdots & -1 & 2  & -1 \\ 
+0 & \cdots & 0 & -1 & 2 \\
+\end{pmatrix}
+\textit{such that }
+\frac{\partial^2U^j}{\partial x^2} \approx A_h*U^j $$
+
+But matrix multiplication is costly, therefore it would be preferable to have the linear operator of the double partial differential instead of the transformation matrix.
+It would look something like:-
+
+    function double_partial(x,dx)
+      for i in 2:length(dx)-1
+        dx[i] = -1*x[i-1] + 2*x[i] + -1*x[i+1]
+      end
+      dx[1] = 2*x[1] + -1*x[2]
+      dx[end] = -1*x[end-1] + 2*x[end]
+    end
+
+So to convert the PDE into an ODE, we discritize the equation in space but not in time. Then this ODE can be solved efficiently by the existing solvers. Our heat equations transforms to the following ODE.
+$u_i' = A_{h}u_i + f(t,u_i)$
+Where A is a linear operator and not the transformation matrix. Thus we will have to make the ODE solvers of DiffEq.jl compatible with linear operators also.
+
+Although it is tedious to compute the Taylor coefficients by hand, Fornberg gave an [algorithm](https://amath.colorado.edu/faculty/fornberg/Docs/MathComp_88_FD_formulas.pdf) to compute them efficiently for any derivative and approximation order. These stencils can efficiently compute derivatives at any point by taking appropriately weighted sums of neighboring points. For example, $[-1, 2, -1]$ is the second order stencil for calculating the 2nd derivative at a point.
+
+Stencil multiplications are **embarrassingly parallel** and this have been taken cared of **DiffEqOperators.jl**.
+
+## Solving the Heat equation using DiffEqOperators
+
+Now let us solve the solve the famous heat equation using the explicit discretization on a 2D `space x time` grid. The heat equation is:-
+        
+$$\frac{\partial u}{\partial t} - \frac{{\partial}^2 u}{\partial x^2} = 0$$
+
+For this example we consider a Dirichlet boundary condition with the initial distribution being parabolic. Since we have fixed the value at boundaries (in this case equal), after a long time we expect the 1D rod to be heated in a linear manner.
+
+        julia> using DiffEqOperators, DifferentialEquations, Plots
+        julia> x = collect(-pi : 2pi/511 : pi);
+        julia> u0 = -(x - 0.5).^2 + 1/12;
+        julia> A = DerivativeOperator{Float64}(2,2,2pi/511,512,:Dirichlet,:Dirichlet;BC=(u0[1],u0[end]));
```

Comment:
Explain this piece of code. Mention that this is for the second order approximation to the second derivative.